### PR TITLE
Add [Criterion] badge

### PR DESCRIPTION
**Pull request by chmoder**
Hi, this badge is for a new service I am implementing for tracking software performance at the function/method level.  It uses [criterion-rs](https://github.com/bheisler/criterion.rs), but I hope that in the future other benchmark frameworks will be added to CI/CD flows.

I am adding the uploader and badge to my projects to accumulate data over time so I can add enhancements.   

Thank you for considering this PR, I look forward to your feedback.

### CODE REVIEW

**Review by PyvesB — COMMENTED** [review]
Hello @chmoder, thanks for your contribution! The implementation mostly looks good to me, however, I've left a few comments and observations inline. 😉

File: services/criterion/criterion.service.js
Comment (calebcartwright) [reply services/criterion/criterion.service.js:8]: agreed, was going to make the same observation

File: services/criterion/criterion.service.js
Comment (calebcartwright) [inline services/criterion/criterion.service.js:44]: We've recently moved to leveraging static fields now that our prod runtime is on a new enough version of Node. We're going to have a lot of existing classes that need to be converted, but let's go ahead and leverage the new paradigm with this new service class

```suggestion
  static category = 'analysis'
  static route = { base: 'criterion', pattern: ':user/:repo' }
  static examples = [
    {
          title: 'Criterion',
          namedParams: {
            user: 'chmoder',
            repo: 'data_vault',
          },
          staticPreview: this.render({ isImproved: true, status: IMPROVED }),
      },
  ]
  
  static defaultBadgeData = { label: 'criterion' }
```

File: services/criterion/criterion.service.js
Comment (calebcartwright) [inline services/criterion/criterion.service.js:47]: bit of a nit, but not clear to me why we're checking the status value twice. my sense is that there's some set of possible values (perhaps as narrow as `improved` and `regressed`), and that from a rendering perspective we want a `brightgreen` color if `status` was `improved`, otherwise `yellow`.

not a blocker for me, but i think this could be simplified as a `if (status === 'improved') { ....` and eliminating the param in the sig here and check and arg within `handle`

File: services/criterion/criterion.tester.js
Comment (chmoder) [reply services/criterion/criterion.tester.js]: You are right, this is strange.  So I just formatted them by hand.  I like the way separate lines reads better anyway.

File: services/criterion/criterion.service.js
Comment (chmoder) [reply services/criterion/criterion.service.js]: Same, i took this pattern from another larger badge and don't need this for this badge.

File: services/criterion/criterion.service.js
Comment (chmoder) [reply services/criterion/criterion.service.js]: Actually this is a great nit.  Since this is my first time in the project I used examples to make this badge.  I am glad you suggested the change.  At this time, there are only two "text" statuses, and 404 response coming from the API.  This is unneeded complexity and removing it, makes it more readable.

File: services/criterion/criterion.service.js
Comment (chmoder) [reply services/criterion/criterion.service.js]: 👍

File: services/criterion/criterion.service.js
Comment (chmoder) [reply services/criterion/criterion.service.js:40]: > Badges should not obtain data from undocumented or reverse-engineered API endpoints.

I see, this is a little unfortunate but also understandable.  I am basically just promoting this idea from from their docs: [Performance has improved.
](https://bheisler.github.io/criterion.rs/book/user_guide/command_line_output.html)

I definitely need and want documentation, so I can make that a priority.  I don't think we should make an exception to the guidelines for my particular badge.

Comment (calebcartwright) [reply services/criterion/criterion.service.js:40]: One of the main drivers for this is the myriad upstream services/tools. that Shields integrates with to provide badges for and the small group of Shields maintainers. We're definitely not experts nor intimately familiar with every service and their respective API contracts, so it's really important that we have things like reference material, docs, contact methods, etc. 

Even something as simple as a Swagger endpoint and an open repo for issues/contact on GitHub/Gitlab/Bitbucket/wherever would be a great starting point.

Comment (chmoder) [reply services/criterion/criterion.service.js:40]: I actually made this [swagger endpoint](https://app.swaggerhub.com/apis-docs/chmoder/Criterion.dev/1.0.0) last night last night.  I was holding off to show you till I added the complete set of available routes; but at this time it has the status route documented.

Right now the repo is private, GitHub has a solution i'm not a fan of, to make an [issues only repo](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/creating-an-issues-only-repository).  I just need to host my authentication information (database connections) somewhere else before I can make it public.

It has been great collaborating with you both, and a good learning experience.  I will follow up on this issue when these things are ready for review.

Comment (chmoder) [reply services/criterion/criterion.service.js:40]: Hello @calebcartwright!  

I decided to open an issues only repo after all because I actually do want to keep the service private.  I'm still glad I moved the sensitive data out of the repository, so that wasn't lost time.  Below are the requested resources, let me know if there is anything else I can do.

Support and contact:
https://github.com/chmoder/api.criterion.dev

API documentation:
https://app.swaggerhub.com/apis-docs/chmoder/Criterion.dev

Comment (PyvesB) [reply services/criterion/criterion.service.js:40]: Looks good to me, let's see what @calebcartwright thinks.

Maybe worth putting the GitHub repo link (which itself links to the API docs) in a comment in the badge code directly?

File: services/criterion/criterion.service.js
Comment (calebcartwright) [inline services/criterion/criterion.service.js:7]: Given the scope of possible statuses it's probably worth making the schema more explicit (unless you anticipate potential additional ones in the near future)
```suggestion
const schema = Joi.string()
  .allow(IMPROVED_STATUS, REGRESSED_STATUS)
  .required()
```

File: services/criterion/criterion.service.js
Comment (calebcartwright) [inline services/criterion/criterion.service.js:5]:
```suggestion
const { IMPROVED_STATUS, NOT_FOUND_STATUS, REGRESSED_STATUS } = require('./constants')
```

**Review by calebcartwright — APPROVED** [review]
LGTM. I think we could tighten up the schema a bit given the additional context, but not strictly necessary for me.

Thanks for the update docs and PR!

**Review by PyvesB — APPROVED** [review]
Looks all good to me @chmoder, I'm going to go ahead and merge this. Thanks for your contribution and friendliness! 👍🏻

**Comment by PyvesB** [comment]
@chmoder if you've got a Twitter account, I can @ you in a small Tweet about this new badge. ;)

**Comment by chmoder** [comment]
> @chmoder if you've got a Twitter account, I can @ you in a small Tweet about this new badge. ;)

That would be awesome!  I have API interfaces to sign up, etc. but am in the middle of making the web app to make registration and management easy.  And of course sweet visualizations, of performance tracking over time.  If you want to wait for that, I can follow up with you, if not go ahead and announce that for me [@chmoder](https://twitter.com/chmoder) there as well.

**Comment by PyvesB** [comment]
Let's wait when, ping me whenever!

**Comment by chmoder** [comment]
Hi @PyvesB, sorry it has been so long!  https://criterion.dev is in an alpha state, meaning we can now sign up, log in, upload reports, view graphs, and even get badge markup from your account.  It is not as pretty as we want, and the visualizations will get more features soon, but either way I am ready for new accounts now.  Thanks for the inspiration!

**Comment by PyvesB** [comment]
As promised: https://twitter.com/Shields_io/status/1310687916256043015 :)

**Pull request merged**
